cache-manager/cache-manager.php

```diff
@@ -0,0 +1,433 @@
+<?php
+/*
+Plugin name: Cache Manager
+Description: Automatically clears the Varnish cache when necessary
+Author: Automattic
+Author URI: http://automattic.com/
+Version: 1.1
+License: GPL version 2 or later - http://www.gnu.org/licenses/old-licenses/gpl-2.0.html
+*/
+
+require_once( __DIR__ . '/api.php' );
+
+class WPCOM_VIP_Cache_Manager {
+	private $ban_urls = array();
+	private $purge_urls = array();
+	private $site_cache_purged = false;
+
+	static public function instance() {
+		static $instance = false;
+		if ( ! $instance )
+			$instance = new WPCOM_VIP_Cache_Manager;
+		return $instance;
+	}
+
+	public function __construct() {
+		// Execute the healthcheck as quickly as possible
+		if ( '/cache-healthcheck?' === $_SERVER['REQUEST_URI'] ) {
+			if ( function_exists( 'newrelic_end_transaction' ) ) {
+				# See: https://docs.newrelic.com/docs/agents/php-agent/configuration/php-agent-api#api-end-txn
+				newrelic_end_transaction( true );
+			}
+			die( 'ok' );
+		}
+
+		add_action( 'init', array( $this, 'init' ) );
+	}
+
+	public function init() {
+		if ( is_super_admin() && isset( $_GET['cm_purge_all'] ) && check_admin_referer( 'manual_purge' ) ) {
+			$this->purge_site_cache();
+			add_action( 'admin_notices' , array( $this, 'manual_purge_message' ) );
+		}
+
+		add_action( 'clean_post_cache', array( $this, 'queue_post_purge' ) );
+		add_action( 'clean_term_cache', array( $this, 'queue_term_purge' ), 10, 2 );
+		add_action( 'switch_theme', array( $this, 'purge_site_cache' ) );
+
+		add_action( 'activity_box_end', array( $this, 'get_manual_purge_link' ), 100 );
+
+		add_action( 'shutdown', array( $this, 'execute_purges' ) );
+	}
+
+	public function get_manual_purge_link() {
+		global $blog_id;
+
+		$url = wp_nonce_url( admin_url( '?cm_purge_all' ), 'manual_purge' );
+
+		$button =  esc_html__( 'Press the button below to force a purge of your entire page cache.' );
+		$button .= '</p><p><span class="button"><a href="' . $url . '"><strong>';
+		$button .= esc_html__( 'Purge Page Cache' );
+		$button .= '</strong></a></span>';
+
+		$nobutton =  esc_html__( 'You do not have permission to purge the cache for the whole site. Please contact your adminstrator.' );
+
+		if ( is_super_admin() ) {
+			echo "<p>$button</p>\n";
+		} else {
+			echo "<p>$nobutton</p>\n";
```

Comment:
Addressed in https://github.com/Automattic/vip-go-mu-plugins/pull/247/commits/14a51e0feb16cee0d9e7a4bf6cdf43f29c46956b